src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizeFilter.cs

```diff
@@ -0,0 +1,58 @@
+// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Linq;
+using System.Security.Claims;
+using System.Threading.Tasks;
+using Microsoft.AspNet.Security;
+using Microsoft.Framework.DependencyInjection;
+
+namespace Microsoft.AspNet.Mvc
+{
+    /// <summary>
+    /// An implementation of <see cref="IAsyncAuthorizationFilter"/>
+    /// </summary>
+    public class AuthorizeFilter : IAsyncAuthorizationFilter
+    {
+        /// <summary>
+        /// Authorize filter for a specific policy
+        /// </summary>
+        /// <param name="policy"></param>
+        public AuthorizeFilter([NotNull] AuthorizationPolicy policy)
+        {
+            Policy = policy;
+        }
+
+        /// <summary>
+        /// Authorization policy to be used
+        /// </summary>
+        public AuthorizationPolicy Policy { get; private set; }
+
+        /// <inheritdoc />
+        public virtual async Task OnAuthorizationAsync([NotNull] AuthorizationContext context)
+        {
+            // Build a ClaimsPrincipal with the Policy's required authentication types
+            if (Policy.ActiveAuthenticationTypes != null && Policy.ActiveAuthenticationTypes.Any())
+            {
+                var results = await context.HttpContext.AuthenticateAsync(Policy.ActiveAuthenticationTypes);
+                context.HttpContext.User = new ClaimsPrincipal(results.Where(r => r.Identity != null).Select(r => r.Identity));
+            }
+
+            // Allow Anonymous skips all authorization
+            if (context.Filters.Any(item => item is IAllowAnonymous))
+            {
+                return;
+            }
+
+            var httpContext = context.HttpContext;
+            var authService = httpContext.RequestServices.GetRequiredService<IAuthorizationService>();
```

Comment:
Instead of calling getRequired service here(i.e within the filter), the pattern that we follow is to have an implementation of ```IFilterFactory``` and create the instance of the filter there. So in this case we would have something like 
```c# 
public class AuthorizeFilterFactory : IFilterFactory

```
and then create the filter in its create method ( the advantage here is that create method takes in an IServiceProvider which can be used to resolve the authorization service. ).